src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -91,34 +111,120 @@ def self.compile_ast(code, compile_options, boltlib_path)
               end
             end
           end
-        else
-          # When we do not need to load bolt we assume there are no Bolt types to resolve in
-          # AST compilation.
-          Puppet[:node_name_value] = compile_options['certname']
-
-          # Use the existing environment with the requested name
-          Puppet::Pal.in_environment(compile_options['environment'],
-                                     envpath: Puppet[:environmentpath],
-                                     facts: compile_options['facts']['values'],
-                                     variables: compile_options['variables']['values']) do |pal|
-            Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+        end
+      end
+      private_class_method :compile_for_project_plan
+
+      def self.compile_for_environment_plan(code, compile_options, boltlib_path)
+        Puppet[:node_name_value] = compile_options['certname']
+
+        # Prior to PE-29443 variables were in a hash. Serialization between ruby/clojure/json did
+        # not preserver hash order necessary for deserialization. The data strucutre is now stored
+        # in a list for moving data and the list is used to construct an ordered ruby hash.
+        plan_variables = if compile_options['variables']['values'].is_a?(Array)
+                      compile_options['variables']['values'].each_with_object({}) do |param_hash, acc|
+                        acc[param_hash.keys.first] = param_hash.values.first
+                      end
+                    else
+                      compile_options['variables']['values']
+                    end
 
+        target_variables = compile_options.dig('target_variables', 'values') || {}
+
+        variables = {
+          variables: plan_variables,
+          target_variables: target_variables,
+        }
+
+        env_conf = {
+          pre_modulepath: boltlib_path,
```

Comment:
This is not a kwarg for in_tmp_enviornment like it is for in_environment. You maybe want something like `Array(boltlib_path) + Array(compile_options['modulepath'])`. Not sure exactly though